File: packages/server/src/protocol.ts
/** Wire-level request/frame types. Spec: spec/03, spec/04, spec/05. */
import type { JsonValue } from "@rayfold/schema";

export interface RequestOp {
  id: number;
  op: string;
  args?: Record<string, unknown>;
  shape?: string;
  vars?: Record<string, JsonValue>;
  key?: string;
  live?: boolean;
  deadline?: number;
  simulate?: boolean;
  /** Omit `$type` where the schema makes it redundant and omit `meta` unless it carries `replay`. Requires a schema-aware client. */
  compact?: boolean;
  /** Conditional write: the version the client last saw of the entity this command targets (spec 03 section 4a). */
  ifVersion?: string | number;
}

export interface RequestMeta {
  client?: string;
  deadline?: number;
  /** W3C trace context (spec 04 section 4); the HTTP transport copies it from the headers of the same names. */
  traceparent?: string;
  tracestate?: string;
}

export interface RequestEnvelope {
  rayfold?: string;
  ops: RequestOp[];
  meta?: RequestMeta;
}

export const PROTOCOL_CODES = [
  "canceled",
  "unknown",
  "invalid_argument",
  "deadline_exceeded",
  "not_found",
  "already_exists",
  "permission_denied",
  "resource_exhausted",
  "failed_precondition",
  "aborted",
  "out_of_range",
  "unimplemented",
  "internal",
  "unavailable",
  "data_loss",
  "unauthenticated",
] as const;
export type ProtocolCode = (typeof PROTOCOL_CODES)[number];
export type ErrorCode = ProtocolCode | "domain";

export interface WireError {
  code: ErrorCode;
  type?: string;
  message: string;
  data?: unknown;
  path?: string;
  retryable?: boolean;
}

export type PatchOp =
  | { set: string; value: Record<string, unknown> }
  | { del: string }
  | { inv: string[] }
  | { invOp: string[] }
  /** Result-scoped (spec 04 section 2b): merge these fields into the plain object at this result path. */
  | { at: string; value: Record<string, unknown> }
  /** Result-scoped: remove these old positions, then insert these elements at these new positions. */
  | { list: string; del?: number[]; ins?: Array<{ at: number; value: unknown }> };

export interface FrameMeta {
  cost?: number;
  cache?: "hit" | "miss" | "stale";
  ms?: number;
  replay?: boolean;
  cursor?: string;
  [k: string]: unknown;
}

export type Frame =
  | { id: number; data: unknown; meta?: FrameMeta; errors?: WireError[]; fin?: boolean }
  | { id: number; ok: unknown; patch?: PatchOp[]; meta?: FrameMeta; errors?: WireError[]; fin: true }
  | { id: number; item: unknown; meta?: FrameMeta }
  | { id: number; patch: PatchOp[]; meta?: FrameMeta }
  | { id: number; at: string; data: unknown; errors?: WireError[] }
  | { id: number; error: WireError; fin: true }
  | { id: number; fin: true }
  | { error: WireError; fin: true };

export const HTTP_STATUS: Record<ErrorCode, number> = {
  invalid_argument: 400,
  failed_precondition: 400,
  out_of_range: 400,
  unauthenticated: 401,
  permission_denied: 403,
  not_found: 404,
  already_exists: 409,
  aborted: 409,
  resource_exhausted: 429,
  canceled: 499,
  unimplemented: 501,
  unavailable: 503,
  deadline_exceeded: 504,
  domain: 422,
  unknown: 500,
  internal: 500,
  data_loss: 500,
};

const RETRYABLE = new Set<ErrorCode>(["unavailable", "deadline_exceeded", "aborted"]);

/** Thrown by resolvers / runtime; converted to a WireError at the frame boundary. */
export class RayfoldError extends Error {
  readonly code: ErrorCode;
  readonly type: string | undefined;
  readonly data: unknown;
  readonly path: string | undefined;
  readonly retryable: boolean;

  constructor(code: ErrorCode, message: string, opts: { type?: string | undefined; data?: unknown; path?: string | undefined; retryable?: boolean | undefined } = {}) {
    super(message);
    this.name = "RayfoldError";
    this.code = code;
    this.type = opts.type;
    this.data = opts.data;
    this.path = opts.path;
    this.retryable = opts.retryable ?? RETRYABLE.has(code);
  }

  toWire(): WireError {
    const w: WireError = { code: this.code, message: this.message };
    if (this.type !== undefined) w.type = this.type;
    if (this.data !== undefined) w.data = this.data;
    if (this.path !== undefined) w.path = this.path;
    if (this.retryable !== RETRYABLE.has(this.code)) w.retryable = this.retryable;
    return w;
  }

  withPath(path: string): RayfoldError {
    return new RayfoldError(this.code, this.message, { type: this.type, data: this.data, path, retryable: this.retryable });
  }

  static domain(type: string, data: unknown, message?: string): RayfoldError {
    return new RayfoldError("domain", message ?? type, { type, data });
  }
}

/**
 * Raised through `ctx.checkVersion` when a conditional command's `ifVersion` does not match.
 * The executor projects `current` through the op's shape so the client can repair its cache without a GET.
 */
export class VersionConflict extends RayfoldError {
  constructor(
    readonly key: string,
    readonly expected: string | number,
    readonly actual: unknown,
    readonly current: unknown,
  ) {
    super("failed_precondition", `${key} is at version ${String(actual)}, not ${String(expected)}`, { type: "VersionConflict", data: { key, expected, actual } });
  }
}

export function toWireError(e: unknown): WireError {
  if (e instanceof RayfoldError) return e.toWire();
  if (e instanceof Error && e.name === "AbortError") return { code: "canceled", message: "Canceled" };
  return { code: "internal", message: "Internal error" };
}

File: kotlin/rayfold-core/src/main/kotlin/dev/rayfold/core/Protocol.kt
package dev.rayfold.core

import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put

/** Protocol error codes (spec/05 section 2). */
enum class Code(val wire: String) {
    CANCELED("canceled"), UNKNOWN("unknown"), INVALID_ARGUMENT("invalid_argument"), DEADLINE_EXCEEDED("deadline_exceeded"),
    NOT_FOUND("not_found"), ALREADY_EXISTS("already_exists"), PERMISSION_DENIED("permission_denied"),
    RESOURCE_EXHAUSTED("resource_exhausted"), FAILED_PRECONDITION("failed_precondition"), ABORTED("aborted"),
    OUT_OF_RANGE("out_of_range"), UNIMPLEMENTED("unimplemented"), INTERNAL("internal"), UNAVAILABLE("unavailable"),
    DATA_LOSS("data_loss"), UNAUTHENTICATED("unauthenticated"), DOMAIN("domain");
}

/** Thrown by resolvers and the runtime; becomes an `error` frame member. */
class RayfoldException(
    val code: Code,
    override val message: String,
    val type: String? = null,
    val data: JsonElement? = null,
    val path: String? = null,
) : RuntimeException(message) {
    fun withPath(p: String) = RayfoldException(code, message, type, data, p)

    fun toWire(): JsonObject = buildJsonObject {
        put("code", code.wire)
        if (type != null) put("type", type)
        put("message", message)
        if (data != null) put("data", data)
        if (path != null) put("path", path)
    }

    companion object {
        fun domain(type: String, data: JsonElement, message: String = type) = RayfoldException(Code.DOMAIN, message, type, data)
        fun of(e: Throwable): RayfoldException = e as? RayfoldException ?: RayfoldException(Code.INTERNAL, "Internal error")
    }
}

/** Raised by [RayfoldContext.checkVersion]; the executor turns it into a VersionConflict carrying the current entity. */
class VersionConflictException(val key: String, val expected: JsonElement, val actual: JsonElement, val current: JsonObject) :
    RuntimeException("$key is at version ${(actual as? JsonPrimitive)?.content ?: actual}, not ${(expected as? JsonPrimitive)?.content ?: expected}")

/** One request op (spec/03 section 1). */
data class RequestOp(
    val id: Int,
    val op: String,
    val args: JsonObject = JsonObject(emptyMap()),
    val shape: String? = null,
    val vars: JsonObject = JsonObject(emptyMap()),
    val key: String? = null,
    val live: Boolean = false,
    val deadline: Long? = null,
    val simulate: Boolean = false,
    val compact: Boolean = false,
    val ifVersion: JsonElement? = null,
) {
    companion object {
        const val MAX_DEADLINE_MS = 600_000L

        fun from(o: JsonObject): RequestOp = RequestOp(
            id = StrictJson.integerOrNull(o["id"])?.toIntOrNull() ?: -1, // "2", 1.9 and 1e12 are not ids
            op = (o["op"] as? JsonPrimitive)?.content ?: "",
            args = o["args"] as? JsonObject ?: JsonObject(emptyMap()),
            shape = (o["shape"] as? JsonPrimitive)?.takeIf { it.isString }?.content,
            vars = o["vars"] as? JsonObject ?: JsonObject(emptyMap()),
            key = (o["key"] as? JsonPrimitive)?.takeIf { it.isString }?.content,
            live = (o["live"] as? JsonPrimitive)?.content == "true",
            deadline = StrictJson.integerOrNull(o["deadline"])?.toLongOrNull()?.takeIf { it in 0..MAX_DEADLINE_MS },
            simulate = (o["simulate"] as? JsonPrimitive)?.content == "true",
            compact = (o["compact"] as? JsonPrimitive)?.content == "true",
            ifVersion = o["ifVersion"]?.takeIf { it !is JsonNull },
        )
    }
}

data class RequestEnvelope(val ops: List<RequestOp>, val meta: JsonObject = JsonObject(emptyMap()), val raw: JsonObject) {
    companion object {
        fun from(o: JsonObject): RequestEnvelope {
            // a non-object entry becomes an empty placeholder; BatchRunner.validate reports it as `ops[i]: expected an object`
            val ops = (o["ops"] as? JsonArray)?.map { RequestOp.from(it as? JsonObject ?: JsonObject(emptyMap())) } ?: emptyList()
            return RequestEnvelope(ops, o["meta"] as? JsonObject ?: JsonObject(emptyMap()), o)
        }
    }
}

/** Frames are plain JSON objects so conformance can compare them structurally (spec/04 section 1). */
object Frames {
    fun data(id: Int, data: JsonElement, cost: Long, errors: List<JsonObject>, fin: Boolean, compact: Boolean = false): JsonObject = buildJsonObject {
        put("id", id); put("data", data)
        if (!compact) put("meta", buildJsonObject { put("cost", cost) })
        if (errors.isNotEmpty()) put("errors", JsonArray(errors))
        if (fin) put("fin", true)
    }
    fun ok(id: Int, ok: JsonElement, patch: List<JsonObject>, cost: Long, errors: List<JsonObject>, replay: Boolean = false, compact: Boolean = false): JsonObject = buildJsonObject {
        put("id", id); put("ok", ok); put("patch", JsonArray(patch))
        if (!compact) put("meta", buildJsonObject { put("cost", cost); if (replay) put("replay", true) })
        if (errors.isNotEmpty()) put("errors", JsonArray(errors))
        put("fin", true)
    }
    /** spec 04 section 2: an `item` frame carries `errors` of its own, as a `data` frame does. */
    fun item(id: Int, item: JsonElement, errors: List<JsonObject> = emptyList()): JsonObject = buildJsonObject {
        put("id", id); put("item", item)
        if (errors.isNotEmpty()) put("errors", JsonArray(errors))
    }
    fun patch(id: Int, patch: List<JsonObject>): JsonObject = buildJsonObject { put("id", id); put("patch", JsonArray(patch)) }
    fun at(id: Int, at: String, data: JsonElement, errors: List<JsonObject>): JsonObject = buildJsonObject {
        put("id", id); put("at", at); put("data", data)
        if (errors.isNotEmpty()) put("errors", JsonArray(errors))
    }
    fun error(id: Int, e: RayfoldException): JsonObject = buildJsonObject { put("id", id); put("error", e.toWire()); put("fin", true) }
    fun fin(id: Int): JsonObject = buildJsonObject { put("id", id); put("fin", true) }
    fun batchError(e: RayfoldException): JsonObject = buildJsonObject { put("error", e.toWire()); put("fin", true) }
}

val NULL: JsonElement = JsonNull
